### ISSUE TITLE
Keynode State Location on Disk

### ISSUE THREAD

**Issue by CMCDragonkai** [MatrixAI/Polykey-Desktop#18]
The keynode state has to be location on disk. By default matching the Unix and Windows user security model, it would be located somewhere under the user's home directory!

For unix: `~/.polykey`.

For windows: `~/AppData/Local/Polykey`.

The creation of the keynode needs to be automatic. Although we can present options similar to GPG, but we should just have sensible defaults in this case.

There will be options in the front such as `polykey --state=./tmp` that will pre-answer those questions. Only when we run `polykey` with no options and it detects nothing in `~/.polykey` will it run those automated questions.

Basically by default, we can use the user home directory for the state directory. But really the state directory can be anywhere. And this in turn works well with agent's ability #15 to be ran anywhere too.

**Comment by CMCDragonkai** [comment]
For mobile devices, the state location will be limited to the app's only filesystem.

**Comment by CMCDragonkai** [comment]
Since this in this repo and not `js-polykey`, this is a concern of the CLI app, and not the core library.

**Comment by MeanMangosteen** [comment]
A minor proposition: I don't think 'state' is the best term to use in this situation. I feel it is a bit abstract, it even wasn't clear to me straight away what you meant by state. We could even make do without the explicit cli option and have something like `polykey init ./tmp`.

**Comment by CMCDragonkai** [comment]
Keynode is another term.

On 16 April 2019 13:23:34 GMT+10:00, ll-aashwin-ll <notifications@github.com> wrote:
>A minor proposition: I don't think 'state' is the best term to use in
>this situation. I feel it is a bit abstract, it even wasn't clear to me
>straight away what you meant by state. We could even make do without
>the explicit cli option and have something like `polykey init ./tmp`.
>
>-- 
>You are receiving this because you authored the thread.
>Reply to this email directly or view it on GitHub:
>https://github.com/MatrixAI/PolyKey/issues/18#issuecomment-483496485

-- 
Sent from my Android device with K-9 Mail. Please excuse my brevity.

**Comment by robert-cronin** [comment]
In the creation of a certain keynode, the cli first asks for location and presents a fs autocomplete interface that starts at the home directory. This happens if there is no keynode location cached in memory ([configstore](https://www.npmjs.com/package/configstore) is currently being used to cache the state)

The node-based applications (e.g. electron) will use `os.homeDir()` to determine the polykey path and suffix with `.polykey` as a default.

Nativscript will pass in sensible defaults to the various constructors.

I think this issue is solved for now.
